src/core/lib/iomgr/executor.cc

```diff
@@ -85,217 +75,260 @@ static size_t run_closures(grpc_closure_list list) {
+bool GrpcExecutor::IsThreaded() {
+  return gpr_atm_no_barrier_load(&num_threads) > 0;
 }
 
-void grpc_executor_set_threading(bool threading) {
-  gpr_atm cur_threads = gpr_atm_no_barrier_load(&g_cur_threads);
+void GrpcExecutor::SetThreading(bool threading) {
+  gpr_atm curr_num_threads = gpr_atm_no_barrier_load(&num_threads);
+
   if (threading) {
-    if (cur_threads > 0) return;
-    g_max_threads = GPR_MAX(1, 2 * gpr_cpu_num_cores());
-    gpr_atm_no_barrier_store(&g_cur_threads, 1);
+    if (curr_num_threads > 0) return;
+
+    // TODO (sreek): max_threads initialization can be moved into the
+    // constructor
+    max_threads = GPR_MAX(1, 2 * gpr_cpu_num_cores());
+    gpr_atm_no_barrier_store(&num_threads, 1);
     gpr_tls_init(&g_this_thread_state);
-    g_thread_state = static_cast<thread_state*>(
-        gpr_zalloc(sizeof(thread_state) * g_max_threads));
-    for (size_t i = 0; i < g_max_threads; i++) {
-      gpr_mu_init(&g_thread_state[i].mu);
-      gpr_cv_init(&g_thread_state[i].cv);
-      g_thread_state[i].thd = grpc_core::Thread();
-      g_thread_state[i].elems = GRPC_CLOSURE_LIST_INIT;
+    thd_state = static_cast<thread_state*>(
+        gpr_zalloc(sizeof(thread_state) * max_threads));
+
+    for (size_t i = 0; i < max_threads; i++) {
+      gpr_mu_init(&thd_state[i].mu);
+      gpr_cv_init(&thd_state[i].cv);
+      thd_state[i].id = i;
+      thd_state[i].thd = grpc_core::Thread();
+      thd_state[i].elems = GRPC_CLOSURE_LIST_INIT;
     }
 
-    g_thread_state[0].thd =
-        grpc_core::Thread("grpc_executor", executor_thread, &g_thread_state[0]);
-    g_thread_state[0].thd.Start();
+    thd_state[0].thd =
+        grpc_core::Thread(name, &GrpcExecutor::ThreadMain, &thd_state[0]);
+    thd_state[0].thd.Start();
   } else {
-    if (cur_threads == 0) return;
-    for (size_t i = 0; i < g_max_threads; i++) {
-      gpr_mu_lock(&g_thread_state[i].mu);
-      g_thread_state[i].shutdown = true;
-      gpr_cv_signal(&g_thread_state[i].cv);
-      gpr_mu_unlock(&g_thread_state[i].mu);
+    if (curr_num_threads == 0) return;
+
+    for (size_t i = 0; i < max_threads; i++) {
+      gpr_mu_lock(&thd_state[i].mu);
+      thd_state[i].shutdown = true;
+      gpr_cv_signal(&thd_state[i].cv);
+      gpr_mu_unlock(&thd_state[i].mu);
     }
-    /* ensure no thread is adding a new thread... once this is past, then
-       no thread will try to add a new one either (since shutdown is true) */
-    gpr_spinlock_lock(&g_adding_thread_lock);
-    gpr_spinlock_unlock(&g_adding_thread_lock);
-    for (gpr_atm i = 0; i < g_cur_threads; i++) {
-      g_thread_state[i].thd.Join();
+
+    /* Ensure no thread is adding a new thread. Once this is past, then no
+     * thread will try to add a new one either (since shutdown is true) */
+    gpr_spinlock_lock(&adding_thread_lock);
+    gpr_spinlock_unlock(&adding_thread_lock);
+
+    for (gpr_atm i = 0; i < num_threads; i++) {
+      thd_state[i].thd.Join();
```

Comment:
Given this is a possible point where we may block forever, I suggest adding a tracer logging message after each successful `Join()`. That way we could quickly see if some thread is refusing to terminate.

Reply:
good suggestion.  Added the trace